File: android/src/main/kotlin/com/amplitude/experiment/flutter/ExperimentSdkCodec.kt
package com.amplitude.experiment.flutter


import com.amplitude.experiment.flutter.ExperimentConfigData as FlutterConfig
import com.amplitude.experiment.ExperimentConfig
import com.amplitude.experiment.flutter.Variant as FlutterVariant
import com.amplitude.experiment.Variant
import com.amplitude.experiment.flutter.ExperimentUser as FlutterExperimentUser
import com.amplitude.experiment.ExperimentUser
import com.amplitude.experiment.flutter.Exposure as FlutterExposure
import com.amplitude.experiment.Exposure
import com.amplitude.experiment.ExposureTrackingProvider
import com.amplitude.experiment.FetchOptions as SdkFetchOptions
import com.amplitude.experiment.flutter.FetchOptions as FlutterFetchOptions

fun convertUser(flutterUser: FlutterExperimentUser?): ExperimentUser? {
    return flutterUser?.let {
        val builder = ExperimentUser.Builder()
         builder.deviceId(it.deviceId)
         builder.userId(it.userId)
         builder.country(it.country)
         builder.city(it.city)
         builder.region(it.region)
         builder.dma(it.dma)
         builder.language(it.language)
         builder.platform(it.platform)
         builder.version(it.version)
         builder.os(it.os)
         builder.deviceModel(it.deviceModel)
         builder.deviceBrand(it.deviceBrand)
         builder.deviceManufacturer(it.deviceManufacturer)
         builder.carrier(it.carrier)
         builder.library(it.library)

        builder.userProperties(it.userProperties)
        builder.groups(groupsFromPigeon(it.groups))
        builder.groupProperties(it.groupProperties)
        builder.build()
    }
}
fun convertUser(user: ExperimentUser): FlutterExperimentUser {
    return FlutterExperimentUser(deviceId = user.deviceId,
    userId = user.userId,
    country = user.country,
    city = user.city,
    region = user.region,
    dma = user.dma,
    language = user.language,
    platform = user.platform,
    version = user.version,
    os = user.os,
    deviceModel = user.deviceModel,
    deviceBrand = user.deviceBrand,
    deviceManufacturer = user.deviceManufacturer,
    carrier = user.carrier,
    library = user.library,
    userProperties = user.userProperties,
    groups = groupsToPigeon(user.groups),
    groupProperties = user.groupProperties)
}

fun convertConfig(flutterConfig: FlutterConfig, api: CustomProviderApi): ExperimentConfig {
    val builder = ExperimentConfig.builder()
    builder.instanceName(flutterConfig.instanceName)
    builder.initialFlags(flutterConfig.initialFlags)
    builder.initialVariants(variantsFromPigeon(flutterConfig.initialVariants))
    builder.source(parseSource(flutterConfig.source))
    builder.serverZone(parseServerZone(flutterConfig.serverZone))
    builder.serverUrl(flutterConfig.serverUrl)
    builder.flagsServerUrl(flutterConfig.flagsServerUrl)
    builder.fetchTimeoutMillis(flutterConfig.fetchTimeoutMillis)
    builder.retryFetchOnFailure(flutterConfig.retryFetchOnFailure)
    builder.automaticExposureTracking(flutterConfig.automaticExposureTracking)
    builder.fetchOnStart(flutterConfig.fetchOnStart)
    builder.pollOnStart(flutterConfig.pollOnStart)
    builder.automaticFetchOnAmplitudeIdentityChange(flutterConfig.automaticFetchOnAmplitudeIdentityChange)
    if (flutterConfig.hasTrackingProvider) {
        builder.exposureTrackingProvider(generateExposureTrackingProvider(flutterConfig.instanceName, api))
    }
    return builder.build()
}

fun convertFetchOptions(options: FlutterFetchOptions?): SdkFetchOptions? {
    return options?.let { SdkFetchOptions(it.flagKeys) }
}

fun convertExposure(exposure: Exposure): FlutterExposure {
    return FlutterExposure(exposure.flagKey, exposure.variant, exposure.experimentKey, exposure.metadata)
}

fun variantFromPigeon(flutterVariant: FlutterVariant?): Variant? {
    return flutterVariant?.let { Variant(it.value, it.payload, it.expKey, it.key, it.metadata)}
}

fun variantToPigeon(variant: Variant): FlutterVariant {
    return FlutterVariant(variant.key, variant.value, jsonToStandard(variant.payload), variant.expKey, jsonToMap(variant.metadata))
}

fun variantsToPigeon(variants: Map<String, Variant>): Map<String, FlutterVariant> {
    return variants.mapValues { (_, v) -> variantToPigeon(v) }
}

private fun variantsFromPigeon(flutterVariants :Map<String, FlutterVariant>): Map<String, Variant> {
    return flutterVariants.mapValues { (_, fv) -> variantFromPigeon(fv)!! }
}

private fun groupsFromPigeon(flutterGroup: Map<String, List<String>>?) :Map<String, Set<String>>? {
    return flutterGroup?.let { it.mapValues { (_, value) -> value.toSet() } }
}

private fun groupsToPigeon(group: Map<String, Set<String>>?): Map<String, List<String>>? {
    return group?.let { it.mapValues { (_, value) -> value.toList() }}
}

private fun parseSource(flutterSource: Source): com.amplitude.experiment.Source {
    return when (flutterSource) {
        Source.LOCAL_STORAGE -> com.amplitude.experiment.Source.LOCAL_STORAGE
        Source.INITIAL_VARIANTS -> com.amplitude.experiment.Source.INITIAL_VARIANTS
    }
}

private fun parseServerZone(flutterServerZone: ServerZone): com.amplitude.experiment.ServerZone {
    return when (flutterServerZone) {
        ServerZone.US -> com.amplitude.experiment.ServerZone.US
        ServerZone.EU -> com.amplitude.experiment.ServerZone.EU
    }
}

private fun jsonToMap(value: Any?): Map<String, Any?>? {
    if (value == null) return null
    if (value is Map<*, *>) return value as Map<String, Any?>
    if (value is org.json.JSONObject) {
        val map = mutableMapOf<String, Any?>()
        value.keys().forEach { key ->
            map[key] = jsonToStandard(value.get(key))
        }
        return map
    }
    return null
}

private fun jsonToStandard(value: Any?): Any? {
    return when (value) {
        null, org.json.JSONObject.NULL -> null
        is org.json.JSONObject -> jsonToMap(value)
        is org.json.JSONArray -> (0 until value.length()).map { jsonToStandard(value.get(it)) }
        else -> value  // String, Int, Double, Boolean — already supported
    }
}

private fun generateExposureTrackingProvider(instanceName: String, api: CustomProviderApi): ExposureTrackingProvider {
    return object : ExposureTrackingProvider {
        override fun track(exposure: Exposure) {
            api.track(
                instanceName,convertExposure(exposure),
            ) {}
        }
    }
}

File: ios/Classes/ExperimentSdkCodec.swift
import AmplitudeExperiment
import Foundation

/**
 * Converts between Pigeon/Flutter types (from AmplitudeExperimentApi.g.swift) and
 * Amplitude Experiment SDK types. Mirrors the Android ExperimentSdkCodec.kt.
 */
enum ExperimentSdkCodec {
    /// Convert Pigeon ExperimentUser to SDK ExperimentUser.
    static func convertUser(_ pigeon: ExperimentUser?) -> AmplitudeExperiment.ExperimentUser? {
        guard let pigeon = pigeon else { return nil }
        let builder = AmplitudeExperiment.ExperimentUserBuilder()
        builder.deviceId(pigeon.deviceId)
        builder.userId(pigeon.userId)
        builder.country(pigeon.country)
        builder.city(pigeon.city)
        builder.region(pigeon.region)
        builder.dma(pigeon.dma)
        builder.language(pigeon.language)
        builder.platform(pigeon.platform)
        builder.version(pigeon.version)
        builder.os(pigeon.os)
        builder.deviceModel(pigeon.deviceModel)
        builder.deviceManufacturer(pigeon.deviceManufacturer)
        builder.carrier(pigeon.carrier)
        builder.library(pigeon.library)
        builder.userProperties(pigeon.userProperties)
        builder.groups(pigeon.groups)
        builder.groupProperties(pigeon.groupProperties)
        return builder.build()
    }

    /// Convert SDK ExperimentUser to Pigeon ExperimentUser (for getUser return value).
    static func convertUserToPigeon(_ sdk: AmplitudeExperiment.ExperimentUser?) -> ExperimentUser {
        guard let sdk = sdk else {
            return ExperimentUser()
        }
        var groups: [String: [String]]?
        if let sdkGroups = sdk.groups {
            let pairs: [(String, [String])] = sdkGroups.map { ($0.key, Array($0.value)) }
            groups = Dictionary(uniqueKeysWithValues: pairs)
        }
        return ExperimentUser(
            deviceId: sdk.deviceId,
            userId: sdk.userId,
            country: sdk.country,
            city: sdk.city,
            region: sdk.region,
            dma: sdk.dma,
            language: sdk.language,
            platform: sdk.platform,
            version: sdk.version,
            os: sdk.os,
            deviceModel: sdk.deviceModel,
            deviceBrand: nil,
            deviceManufacturer: sdk.deviceManufacturer,
            carrier: sdk.carrier,
            library: sdk.library,
            ipAddress: nil,
            userProperties: sdk.userProperties,
            groups: groups,
            groupProperties: sdk.groupProperties
        )
    }

    /// Convert Pigeon ExperimentConfigData to SDK ExperimentConfig.
    static func convertConfig(_ pigeon: ExperimentConfigData, api: CustomProviderApi) -> AmplitudeExperiment.ExperimentConfig {
        let builder = AmplitudeExperiment.ExperimentConfigBuilder()
        builder.instanceName(pigeon.instanceName)
        if let fallback = convertVariant(pigeon.fallbackVariant) {
            builder.fallbackVariant(fallback)
        }
        builder.initialFlags(pigeon.initialFlags)
        builder.initialVariants(convertVariants(pigeon.initialVariants))
        builder.source(convertSource(pigeon.source))
        builder.serverZone(convertServerZone(pigeon.serverZone))
        builder.serverUrl(pigeon.serverUrl)
        builder.flagsServerUrl(pigeon.flagsServerUrl)
        builder.fetchTimeoutMillis(Int(pigeon.fetchTimeoutMillis))
        builder.fetchRetryOnFailure(pigeon.retryFetchOnFailure)
        builder.automaticExposureTracking(pigeon.automaticExposureTracking)
        builder.fetchOnStart(pigeon.fetchOnStart)
        builder.pollOnStart(pigeon.pollOnStart)
        builder.automaticFetchOnAmplitudeIdentityChange(pigeon.automaticFetchOnAmplitudeIdentityChange)
        if pigeon.hasTrackingProvider {
            builder.exposureTrackingProvider(
                PigeonExposureTrackingProvider(instanceName: pigeon.instanceName, api: api)
            )
        }
        return builder.build()
    }

    /// Convert Pigeon Variant to SDK Variant.
    static func convertVariant(_ pigeon: Variant?) -> AmplitudeExperiment.Variant? {
        guard let pigeon = pigeon else { return nil }
        return AmplitudeExperiment.Variant(
            pigeon.value,
            payload: pigeon.payload,
            expKey: pigeon.expKey,
            key: pigeon.key,
            metadata: pigeon.metadata
        )
    }

    /// Convert SDK Variant to Pigeon Variant (for return values).
    static func convertVariant(_ sdk: AmplitudeExperiment.Variant) -> Variant {
        Variant(
            key: sdk.key,
            value: sdk.value,
            payload: sdk.payload,
            expKey: sdk.expKey,
            metadata: sdk.metadata
        )
    }

    /// Convert SDK Exposure to Pigeon Exposure (for tracking provider callback).
    static func convertExposure(_ sdk: AmplitudeExperiment.Exposure) -> Exposure {
        Exposure(
            flagKey: sdk.flagKey,
            variant: sdk.variant,
            experimentKey: sdk.experimentKey,
            metadata: sdk.metadata
        )
    }

    /// Convert Pigeon FetchOptions to SDK FetchOptions.
    static func convertFetchOptions(_ pigeon: FetchOptions?) -> AmplitudeExperiment.FetchOptions? {
        guard let pigeon = pigeon else { return nil }
        return AmplitudeExperiment.FetchOptions(pigeon.flagKeys)
    }

    private static func convertVariants(_ pigeon: [String: Variant]) -> [String: AmplitudeExperiment.Variant] {
        pigeon.mapValues { convertVariant($0)! }
    }

    private static func convertSource(_ pigeon: Source) -> AmplitudeExperiment.Source {
        switch pigeon {
        case .localStorage: return AmplitudeExperiment.Source.LocalStorage
        case .initialVariants: return AmplitudeExperiment.Source.InitialVariants
        }
    }

    private static func convertServerZone(_ pigeon: ServerZone) -> AmplitudeExperiment.ServerZone {
        switch pigeon {
        case .us: return AmplitudeExperiment.ServerZone.US
        case .eu: return AmplitudeExperiment.ServerZone.EU
        }
    }
}

/// Bridges native ExposureTrackingProvider to Dart via Pigeon's CustomProviderApi.
class PigeonExposureTrackingProvider: NSObject, AmplitudeExperiment.ExposureTrackingProvider {
    private let instanceName: String
    private let api: CustomProviderApi

    init(instanceName: String, api: CustomProviderApi) {
        self.instanceName = instanceName
        self.api = api
    }

    func track(exposure: AmplitudeExperiment.Exposure) {
        let pigeonExposure = ExperimentSdkCodec.convertExposure(exposure)
        api.track(instanceName: instanceName, exposure: pigeonExposure) { _ in }
    }
}
